File: Bit_Manipulation/python/flip_bit_to_win.py
from numpy import uint32


def get_bit(num, i):
    return ((num & (1 << i)) != 0)


def set_bit(num, i):
    return num | (1 << i)


def get_ones_count(num):
    count, length = 0, 0

    for i in range(32):
        if get_bit(num, i):
            count += 1
        else:
            if count > length:
                length = count
            count = 0

    return length


def flip(num):
    length = 0
    for i in range(32):
        if not get_bit(num, i):
            flipped = set_bit(num, i)

            ones = get_ones_count(flipped)

            if ones > length:
                length = ones

    return length



def main():
    num = uint32(383)
    print("{0:b}".format(num))

    print(flip(num))


if __name__ == '__main__':
    main()

File: Bit_Manipulation/go/flip_bit_to_win.go
package main

import (
  "fmt"
)


func getBit(num uint32, i int) bool {
  return ((num & (1 << i)) != 0)
}


func setBit(num uint32, i int) uint32 {
  return num | (1 << i)
}


func getOnesCount(num uint32) int {
  var count, length int

  for i := 0; i < 32; i++ {
    if getBit(num, i) {
      count++
    } else {
      if count > length {
        length = count
      }

      count = 0
    }
  }

  return length
}


func flip(num uint32) int {
  var length, ones int
  var flippedNum uint32

  for i := 0; i < 32; i++ {
    if !getBit(num, i) {
      flippedNum = setBit(num, i)

      ones = getOnesCount(flippedNum)
      if ones > length {
        length = ones
      }
    }
  }
  return length
}


func main() {
  var a uint32
  a = 383
  fmt.Printf("%032b\n", a)

  fmt.Println(flip(a))
}
